### PR TITLE
Multi window context render

### PR DESCRIPTION
**Pull request by paul-veevers**
**What**:
Rendering into a different window context, e.g. an iframe. See https://github.com/paypal/downshift/issues/200

**Why**:
Clicks on items don't register.

**How**:
Accept an `eventTarget` prop that defaults to `window`

**Checklist**:
- [x] Documentation
- [ ] Tests
- [ ] Ready to be merged
- [x] Added myself to contributors table

Notes:
- Any ideas on how to test this in a nice way?
- I haven't touched the `set-a11y-status` file. Do you want me to change that as well? As it stands, the wrong `document` will be rendered into.
- Are you ok with the prop name? I couldn't think of anything both more flexible and descriptive.

ps. something that stumped me at the beginning was unresolved import errors from eslint. Turns out I needed to `npm i` from inside the `stories` folder.

### CODE REVIEW

File: src/downshift.js [hunk 1]
```diff
@@ -148,7 +153,7 @@ class Downshift extends Component {
   }
 
   getItemNodeFromIndex = index => {
-    return document.getElementById(this.getItemId(index))
+    return this._rootNode.querySelector(`#${this.getItemId(index)}`)
```
Comment (kentcdodds) [inline src/downshift.js]: This won't work for folks rendering the menu outside the autocomplete root node I'm afraid. Need to set this as the eventTarget. Due to that fact, maybe eventTarget is not quite the right name... Maybe simply "window" ?

Comment (tansongyang) [reply src/downshift.js]: We could also consider "context", since the new docs say that the type can be `Element`/`document`/`window`.

Comment (kentcdodds) [reply src/downshift.js]: I'd be fine with that too, tough I hesitate because I worry folks might confuse that with React's context feature 🤔

Comment (tansongyang) [reply src/downshift.js]: Agreed. I realized right after I typed it that React also has "context".

Looked at the [thesaurus entry for "context"](http://www.thesaurus.com/browse/context). I'll throw out "environment".

I'll also suggest "renderContext".

Comment (kentcdodds) [reply src/downshift.js]: I'm good with either of those. Like the docs said, most people shouldn't be using this prop anyway.

Comment (paul-veevers) [reply src/downshift.js]: "environment" works well for other use cases like React Native (https://github.com/paypal/downshift/issues/185)

**Comment by paul-veevers** [comment]
I've made it so `environment` is must be a `window` object. It makes assumptions in the code much easier, rather than having to do some ternary magic depending on whether it's a window or document or element.

**Comment by tansongyang** [comment]
Are you comfortable with TypeScript? If so, would you mind also updating `typings/index.d.ts`?

**Comment by paul-veevers** [comment]
Definitely not comfortable, but I gave it a go. Let me know if I got anything wrong.

File: typings/index.d.ts [hunk 2]
```diff
@@ -26,6 +26,17 @@ export interface DownshiftProps {
   inputValue?: string
   highlightedIndex?: number
   id?: string
+  environment?: Environment
+}
+
+export interface Environment {
+  addEventListener: (type: string, listener: (event: Event) => void) => void
+  removeEventListener: (type: string, listener: () => void) => void
```
Comment (tansongyang) [inline typings/index.d.ts]: For both `addEventListener` and `removeEventListener`, I suggest just using `Function`  as the callback type (as the rest of this file does). This gives us the most flexibility.

File: typings/index.d.ts [hunk 3]
```diff
@@ -26,6 +26,17 @@ export interface DownshiftProps {
...
+  document: Document
+}
+
+export interface Document {
+  getElementById: (id: String) => HTMLElement
```
Comment (tansongyang) [inline typings/index.d.ts]: Use the lowercase `string` here instead of `String`.

**Comment by paul-veevers** [comment]
Thanks for the comments @tansongyang !

**Comment by tansongyang** [comment]
@paul-veevers Thanks for the hard work!

**Comment by bliitzkrieg** [comment]
This commit breaks SSR because window does not exist in node

**Comment by tansongyang** [comment]
@bliitzkrieg What would you recommend? Is there a way to get this functionality while maintaining support for SSR? Also, are there any automated tests we could add that would catch this in the future? (I know very little about SSR.)

**Comment by bliitzkrieg** [comment]
I believe the fix can be:

```
  static defaultProps = {
    defaultHighlightedIndex: null,
    defaultSelectedItem: null,
    defaultInputValue: '',
    defaultIsOpen: false,
    getA11yStatusMessage,
    id: generateId('downshift'),
    itemToString: i => (i == null ? '' : String(i)),
    onStateChange: () => {},
    onUserAction: () => {},
    onChange: () => {},
    environment: typeof window === 'undefined' ? {} : window,
  }
```

I tested locally and the tests all pass and my app renders on the server fine.

**Comment by kentcdodds** [comment]
Your suggestion works fine for me.

I'll work on the SSR issue. I have some ideas. :+1:

**Comment by kentcdodds** [comment]
See https://github.com/paypal/downshift/pull/205 :)

**Comment by paul-veevers** [comment]
Thank you @kentcdodds and @tansongyang for making my first OSS PR a great experience :)

**Comment by kentcdodds** [comment]
Sure thing! I'm glad you were able to start here!

**Comment by bliitzkrieg** [comment]
Thanks @kentcdodds for looking into it so quickly! I really appreciate the hard work

**Pull request merged**